sled-agent/src/storage_manager.rs

```diff
@@ -955,84 +1030,160 @@ impl StorageWorker {
         loop {
             tokio::select! {
                 _ = self.nexus_notifications.next(),
                     if !self.nexus_notifications.is_empty() => {},
                 Some(request) = self.rx.recv() => {
-                    use StorageWorkerRequest::*;
-                    match request {
-                        AddDisk(disk) => {
-                            self.upsert_disk(
-                                &resources,
-                                disk,
-                                queued_u2_drives,
-                            )
-                            .await?;
-                        },
-                        AddSyntheticDisk(zpool_name) => {
-                            self.upsert_synthetic_disk(
-                                &resources,
-                                zpool_name,
-                                queued_u2_drives,
-                            ).await?;
-                        },
-                        RemoveDisk(disk) => {
-                            self.delete_disk(&resources, disk).await?;
-                        },
-                        NewFilesystem(request) => {
-                            let result = self.add_dataset(&resources, &request).await;
-                            let _ = request.responder.send(result);
-                        },
-                        DisksChanged(disks) => {
-                            self.ensure_using_exactly_these_disks(
-                                &resources,
-                                disks,
-                                queued_u2_drives
-                            )
-                            .await?;
-                        },
-                        SetupUnderlayAccess(UnderlayRequest { underlay, responder }) => {
-                            // If this is the first time establishing an
-                            // underlay we should notify nexus of all existing
-                            // disks and zpools.
-                            //
-                            // Instead of individual notifications, we should
-                            // send a bulk notification as described in https://
-                            // github.com/oxidecomputer/omicron/issues/1917
-                            if self.underlay.lock().await.replace(underlay).is_none() {
-                                self.notify_nexus_about_existing_resources(&resources).await?;
-                            }
-                            let _ = responder.send(Ok(()));
+                    // We want to queue failed requests related to the key manager
+                    match self.handle_storage_worker_request(
+                        resources, queued_u2_drives, request)
+                    .await {
+                        Err(Error::DiskError(_)) => {
+                            // We already handle and log disk errors, no need to
+                            // return here.
```

Comment:
Sorry for a vague complaint, but: this seems hard to verify by reading `handle_storage_worker_request`. Does every function we call in every arm of `handle_store_worker_request` handle every possible `DiskError(_)`? Even if yes today, is there some way to restructure this that makes it easier to ensure it continues to be the case? Maybe we carve `DiskError` out of `Error` entirely? (Not at all sure that's reasonable, just throwing out an idea.)

Reply:
It's a *very* justifiable complaint. I had the same thought when writing it. However, if you take a close look you'll realize that we don't actually do anything with the `DiskErrors` anyway. We just log them and move on in [`do_work`](https://github.com/oxidecomputer/omicron/blob/main/sled-agent/src/storage_manager.rs#L942-L946) . This code doesn't make things any worse.

As John and I discussed offline there are other problems here in that we block waiting for rack secrets to unlock and so we'll need to add more concurrency. Overall, this is more of a general "we need to make sled-agent" a bit more cohesive and testable. We can tackle that outside this PR.